### DISCUSSION TITLE
Is the `hasInitialValue` check of `useHydrateAtoms` necessary?

### DISCUSSION THREAD

**Discussion by jzhang026** [pmndrs/jotai#1691] [General] [answered]
I have a doubt regarding the `hasInitialValue` check of `useHydrateAtoms`:
below is the code of function `restoreAtoms` used during atom hydration, and I also paste the [source code link](https://github.com/pmndrs/jotai/blob/main/src/core/store.ts#L942) here in case you need to refer to it.
```ts
const restoreAtoms = (
    values: Iterable<readonly [AnyAtom, AnyAtomValue]>,
    version?: VersionObject
  ): void => {
    for (const [atom, value] of values) {
      if (hasInitialValue(atom)) {
        setAtomPromiseOrValue(version, atom, value)
        invalidateDependents(version, atom)
      }
    }
    flushPending(version)
  }
```

If I have an async atom like below, then this `postData` atom don't have the `init` property available. (there is no initial state)

```ts
const postData = atom(async (get) => {
  const id = get(postId);
  const { data } = await fetch(`https://hacker-news.firebaseio.com/v0/item/${id}.json`);

  return data;
});
```

thus below atom hydration will be skipped on browser side.
```ts
useHydrateAtoms([
  [postData, dataFromServer]
]);
```

To make it work, I tried to work around it by tweak the hydration code a little bit , like:
```ts
function myUseHydrateAtoms(values) {
  function patch(atom: any) {
    if (!('init' in atom)) {
      atom['init'] = '__FAKE_INIT__';
      return () => delete atom['init'];
    }
    return () => {};
  }
  
  const cleanups = [];
  for (const [atom] of values) {
    cleanups.push(patch(atom));
  }

  // then call the Jotail hydration util
  useHydrateAtoms(values);

  // cleanup the patch after hydration
  cleanups.forEach((cleanup) => cleanup());
}
```

Despite the patch works, I still feel worried and try to understand it clearly. 
Is the `hasInitialValue` check of `useHydrateAtoms` necessary?

**Comment by dai-shi** [comment, marked as answer]
That's necessary to follow the React rule. https://gist.github.com/sebmarkbage/75f0838967cd003cd7f9ab938eb1958f
It will be important in concurrent rendering, but probably harmless without it in legacy rendering.
